Set name record from font.info

moyogo opened the pull request:
- Set name record from font.info, these have priority over the ones built from other info. If they are in the UFO we can assume they should be used directly.
- Add option to set legacy Mac names. Some legacy environments or odd applications need this.
- That option also adds legacy Mac cmap for the same reasons.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlei18n/ufo2ft/66)
<!-- Reviewable:end -->

khaledhosny commented:
Wouldn’t it be simpler to just let the clients that need legacy Mac names set openTypeNameRecords to their liking, that would avoid having to deal with the legacy encoding id and lang id mess of Mac names (which this code does not attempt to handle at all which would produce invalid strings for almost anything outside ASCII).

khaledhosny commented:
Also, I’m curious what apps still require Mac names?

moyogo commented:
> Wouldn’t it be simpler to just let the clients that need legacy Mac names set openTypeNameRecords to their liking, that would avoid having to deal with the legacy encoding id and lang id mess of Mac names (which this code does not attempt to handle at all which would produce invalid strings for almost anything outside ASCII).

That’s exactly why it sets the names from openTypeNameRecords first. Once set they are not overwritten.

moyogo commented:
> Also, I’m curious what apps still require Mac names?

Mac Office apparently.

anthrotype commented:
the --mac-legacy option is off by default, you don't need to use it if you don't want the basic MacRoman English names to be auto-generated.

khaledhosny commented:
> That’s exactly why it sets the names from openTypeNameRecords first. Once set they are not overwritten.

That is my point, the macLegacy attribute is basically redundant here, and it makes the code harder to follow.

khaledhosny commented:
> Mac Office apparently.

Does it also need Mac cmap subtables? (just curios).

anthrotype commented:
> Platform and encoding ID's in the name table should be consistent with those in the cmap table. If they are not, the font will not load in Windows

https://www.microsoft.com/typography/otspec/recom.htm

khaledhosny commented:
Hmm, we have a bigger problem then. If we allow arbitrary encoding ids in the name table (through openTypeNameRecords) then the cmap table needs to account for that (though I wouldn’t believe the wording of the spec here without actually testing it, it often has such statements that have not been true for decades).

anthrotype commented on Lib/ufo2ft/outlineOTF.py:
```diff
@@ -380,6 +404,21 @@ def setupTable_cmap(self):
+            cmap6_1_0.platformID = 1
+            cmap6_1_0.platEncID = 0
+            cmap6_1_0.language = 0
+            # builde mac legacy mapping
```
s/builde/build/

moyogo commented:
TBH we’re not fond of the mac legacy names and cmap. However we have had clients inquire about legacy support and it’s a company policy to support it. Considering we cannot test all the environments that supposedly require mac legacy, we add it.
We’d rather have it as an option than have to postprocess.

I cleaned up the commits so one only adds the openTypeNameRecords and the others deal with mac legacy, for cherry-picking.

Regarding characters being invalid in mac legacy names, there’s a similar issue in CFF trademark and copyright. There’s some normalization for those, but invalid characters can still go through and raise.

khaledhosny commented:
Can we at least force MacRoman encoding for those strings with `ignore` or `replace` error handling? Or does FontTools handle this already? (I guess we can do something similar for CFF strings).

anthrotype commented:
ping @jamesgk @behdad

jamesgk commented on Lib/ufo2ft/outlineOTF.py:
```diff
@@ -380,6 +406,21 @@ def setupTable_cmap(self):
             # update tables registry
             cmap.tables = [cmap4_0_3, cmap4_3_1, cmap12_0_4, cmap12_3_10]
+        if self.macLegacy:
+            from fontTools.ttLib.tables._c_m_a_p import cmap_format_6
```
Does anyone know why we have these cmap imports inline? I'd prefer them at the top of the file if possible.

jamesgk commented on Lib/ufo2ft/outlineOTF.py:
```diff
@@ -276,49 +290,61 @@ def setupTable_name(self):
-            if nameIdVal == 6:
-                # postscript font name
-                nameVal = normalizeStringForPostscript(nameVal)
-            rec = NameRecord()
```
We don't need the NameRecord import anymore.

jamesgk commented:
So wait, has anyone confirmed that adding arbitrary name records without adding corresponding cmaps will break a font on Windows? If so, is it bad to support adding such name records? If not, can't we reasonably expect ufo2ft users to use such name records to include legacy Mac name records as @khaledhosny suggested?

behdad commented:
> So wait, has anyone confirmed that adding arbitrary name records without adding corresponding cmaps will break a font on Windows? If so, is it bad to support adding such name records? If not, can't we reasonably expect ufo2ft users to use such name records to include legacy Mac name records as @khaledhosny suggested?

I don't think it breaks anything.  In fact, I'd go as far as claiming: if there is a Unicode cmap, and a Unicode name, adding more stuff wouldn't break anything. (until it does; but let's ignore that for now!)

moyogo commented:
> Can we at least force MacRoman encoding for those strings with ignore or replace error handling? Or does FontTools handle this already? (I guess we can do something similar for CFF strings).

OK. I added a function to normalize mac legacy names and modified the one for postscript strings. Non mac_roman and non ascii characters are in NFKD form or replaced with ? (or both), also composed if the NFC form is in mac_roman, for example:
- anything invalid becomes '?'
- 'e\u0301' become 'é' in mac_roman and 'e?' in ascii
- 'é' becomes 'e?' in ascii
- '…' becomes '...' in ascii.

moyogo commented:
I removed the commits to "add option to set legacy Mac name/cmap". This now only set names from openTypeNameRecords and modifies normalizeStringForPostscript().
